Add option to limit render fps to game tick rate

reaperrr opened the pull request:
We know there are systems out there that struggle to maintain 50, and since our default simulation speed is only 25 anyway, supporting a lower minimum seems like a fair solution to make things a bit easier for players with slow systems.

Min FPS are still clamped to simulation FPS, so gamespeeds of 'Fast' up to 'Fastest' will still result in higher min (render) FPS for now.

Also widened the sliders to make it a bit easier to set specific values despite the increased range.

anvilvapre commented:
The limit is not also enforced in the code?

Can it be that the minimum was once increased to 60 because the main loop in Game.cs required a minimal of this for it's logic to work? I.e. `renderInterval` at line 781 becoming too large?

reaperrr commented:
> The limit is not also enforced in the code?

I guess the PR title is a bit misleading, what I'm actually allowing players to do here is to set a lower `MaxFramerate` cap that the game tries to maintain.
The only enforced real minimum is the 1-renderframe-per-game-tick limit, as far as I know.

> Can it be that the minimum was once increased to 60 because the main loop in Game.cs required a minimal of this for it's logic to work? I.e. renderInterval at line 781 becoming too large?

I'm fairly sure the internal default of 60 for `MaxFramerate` was just set with VSync/60Hz displays in mind, not anything game loop-related.

I don't know for certain, but I believe the old `MinimumValue` of 50 on the sliders was probably based on the assumption that the vast majority of players would never want to set a lower render framerate cap anyway.

pchote commented:
>  I believe the old MinimumValue of 50 on the sliders was probably based on ...

I chose 50 for two reasons: the first is that this corresponds to the tick rate of the `fastest` game speed. The second was that as you get down to framerates near the tick rate, quantization will cause unstable frame pacing (since you can only have an integer number of renders for each tick).

reaperrr commented:
Hm... would keeping the 50 slider minimum but adding a 'Tie framerate to gamespeed' checkbox option that overrides it possibly be a better approach?

anvilvapre commented:
Ping. Would be nice to have in a upcoming release.

reaperrr committed:
Add option to limit render fps to game tick rate

This helps slow systems that struggle to render 2 frames per game tick.

reaperrr commented:
Updated.

I went ahead with the

> Hm... would keeping the 50 slider minimum but adding a 'Tie framerate to gamespeed' checkbox option that overrides it possibly be a better approach?

approach, and I do think we should include this in the upcoming release, considering the perf issues several people on slower machines are reporting.

Removed 'needs +2' label as the PR has changed substantially.

reaperrr commented:
Rebased and updated.

penev92 commented on OpenRA.Mods.Common/Widgets/Logic/Settings/DisplaySettingsLogic.cs:
```diff
@@ -144,12 +145,16 @@ Func<bool> InitPanel(Widget panel)
 				(ds.Mode == WindowMode.Windowed && (origWidthText != windowWidth.Text || origHeightText != windowHeight.Text));
 
+			var frameLimitGamespeedCheckbox = panel.Get<CheckboxWidget>("FRAME_LIMIT_GAMESPEED_CHECKBOX");
+			var frameLimitGamespeedLabel = frameLimitGamespeedCheckbox.Text;
```
This is unused.

penev92 reviewed (dismissed):
Works as advertised.
👍  with that unused variable removed.

anvilvapre commented:
looking quickly at your changes, 

this does not seem to remove the forced renders. it's these forced renders that cause the issue, as they are forced contineously once the game becomes slow. they won't allow the the game to skip frames to catch up.

reaperrr commented:
> looking quickly at your changes, this does not seem to remove the forced renders.

That's actually not been the goal of this PR.

What this PR does is allowing a minimum framerate equal to the game simulation framerate, as opposed to the bleed minimum of 50 fps.
At normal gamespeed, this effectively cuts the framerate the engine tries to maintain in half (from 50 to 25), which also up to halves the GPU load.

I don't have the time or knowledge to dig in deeper into the render code.
This is meant to ease the performance issue for slow GPUs for the upcoming release, since I don't expect anyone else to fix the forced-renders issue anytime soon (maybe I'm wrong, but development currently seems rather slow, not that I'm much help in that regard).

PunkPun reviewed (dismissed):
LGTM. Just remove the unused`frameLimitGamespeedLabel`

PunkPun commented:
@reaperrr when do you think we can merge this?

penev92 commented:
I guess we can fixup the PR (I have one small nit mentioned above) ourselves and merge.
I would like for @RoosterDragon to take a look before that though.

RoosterDragon commented:
This seems reasonable to me.

penev92 reviewed (approved):
I removed the unused variable myself and rebased.
I'd say this is good to go now.

penev92 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/fd41716b7a43ab76820d5c061cac5bcd21369985...2724c7c682970927c912b78cd14c0f6d1cd68a97)
